File: server/src/cuminc.js
const app = require('./app')
const path = require('path')
const getFilterCTEs = require('./termdb.filter').getFilterCTEs
const spawn = require('child_process').spawn
const serverconfig = require('./serverconfig')
const readline = require('readline')

export function handle_incidence(genomes) {
	return async (req, res) => {
		app.log(req)
		const q = req.query
		if (q.filter) {
			q.filter = JSON.parse(decodeURIComponent(q.filter))
		}
		try {
			const genome = genomes[q.genome]
			if (!genome) throw 'invalid genome'
			const ds = genome.datasets[q.dslabel]
			if (!ds) throw 'invalid dslabel'
			if (!ds.cohort) throw 'ds.cohort missing'
			const tdb = ds.cohort.termdb
			if (!tdb) throw 'no termdb for this dataset'

			const filter = getFilterCTEs(q.filter, q.ds)
			const data = get_sql_query(q, ds, filter)
			const year_to_events = data.map(d => d.time).join('_')
			const events = data.map(d => d.event).join('_')
			const ci_data = await calculate_cuminc(year_to_events, events)
			//console.log('ci_data:', ci_data)
			const final_data = {}
			if (ci_data == null) {
				console.log('No output from R script')
			} else {
				const case_array = []
				for (let i = 0; i < ci_data.case_time.length; i++) {
					case_array.push([ci_data.case_time[i], ci_data.case_est[i], ci_data.low_case[i], ci_data.up_case[i]])
				}
				final_data.keys = ['time', 'cuminc', 'low', 'high']
				final_data.case = case_array
			}
			res.send(final_data)
		} catch (e) {
			res.send({ error: e.message || e })
			if (e.stack) console.log(e.stack)
		}
	}
}

function get_sql_query(q, ds, filter) {
	const values = filter ? filter.values.slice() : []
	values.push(...[q.grade, q.term_id])

	const sql = `WITH
${filter ? filter.filters + ',' : ''}
event1 AS (
	SELECT sample, MIN(years_to_event) as time, 1 as event
	FROM chronicevents
	WHERE grade >= ?
	  AND grade <= 5
	  AND term_id = ?
	  ${filter ? 'AND sample IN ' + filter.CTEname : ''}
	GROUP BY sample
),
event1samples AS (
	SELECT sample
	FROM event1
),
event0 AS (
	SELECT sample, MAX(years_to_event) as time, 0 as event
	FROM chronicevents
	WHERE grade <= 5 
		AND sample NOT IN event1samples
	  ${filter ? 'AND sample IN ' + filter.CTEname : ''}
	GROUP BY sample
)
SELECT * FROM event1
UNION ALL
SELECT * FROM event0
`
	return ds.cohort.db.connection.prepare(sql).all(values)
}

//function calculate_cuminc(year_to_events, events, groups) {
function calculate_cuminc(year_to_events, events) {
	const ci_data = {}
	return new Promise((resolve, reject) => {
		//const ps = spawn('Rscript', ['server/src/cuminc.R', year_to_events, events, groups]) // Should we define Rscript in serverconfig.json?
		const ps = spawn('Rscript', [path.join(serverconfig.binpath, './server/src/cuminc.R'), year_to_events, events]) // Should we define Rscript in serverconfig.json?
		const rl = readline.createInterface({ input: ps.stdout })

		rl.on('line', line => {
			//console.log("R_line:",line)
			if (line.includes('p_value') == true) {
				const line2 = line.split(':')
				ci_data.pvalue = parseFloat(line2[1].replace('"', '').replace(' ', ''), 10)
			} else if (line.includes('control_time') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.control_time = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			} else if (line.includes('control_est') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.control_est = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			} else if (line.includes('case_time') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.case_time = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			} else if (line.includes('case_est') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.case_est = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			} else if (line.includes('low_control') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.low_control = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			} else if (line.includes('up_control') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.up_control = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			} else if (line.includes('low_case') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.low_case = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			} else if (line.includes('up_case') == true) {
				const char_array = line.split(':')[1].split(',')
				ci_data.up_case = char_array.map(v => parseFloat(v.replace(' ', ''), 10))
			}
		})
		rl.on('close', () => {
			resolve(ci_data)
		})
	})
}

File: server/src/cuminc.R
# Rscript cuminc.R 0.3_0.4_0.1_0.2 1_1_0_0

#library(survival) #for survfit function
library(cmprsk) #for cuminc function
args <- commandArgs(TRUE)

year_to_events <- as.numeric(unlist(strsplit(args[1], split = "_")))
events <- as.integer(unlist(strsplit(args[2], split = "_")))
#groups <- as.integer(unlist(strsplit(args[3], split = "_")))

#print (year_to_events)
#print (events)
#print (groups)

#ci_output <- cuminc(ftime=year_to_events, fstatus=events, group=groups, cencode = 0)
ci_output <- cuminc(ftime=year_to_events, fstatus=events, cencode = 0)

#if (length(ci_output[["Tests"]]) == 6) { # Competing risks present
#  print (paste("p_value:",format(round(ci_output[["Tests"]][3], 6), nsmall = 6)))
#}
#if (length(ci_output[["Tests"]]) == 3) { # No competing risks present
#  print (paste("p_value:",format(round(ci_output[["Tests"]][2], 6), nsmall = 6)))
#}

#print (paste("control_time:",toString(ci_output[[1]]$time)))
#print (paste("control_est:", toString(ci_output[[1]]$est)))

#print ("ci_output[[1]]$var")
#print (ci_output[[1]]$var)

## IMPORTANT !!!! - REMEMBER TO CHANGE the index number of case to 2 from 1 when you implement groups (case vs control)

print (paste("case_time:",toString(ci_output[[1]]$time)))
print (paste("case_est:", toString(ci_output[[1]]$est)))

#print ("ci_output[[2]]$var")
#print (ci_output[[2]]$var)

### 95% CI, all negative CI values are set to zero
#low_control <- ci_output[[1]]$est-1.96*sqrt(ci_output[[1]]$var)
#low_control [low_control < 0]  <- 0
#print (paste("low_control:", toString(low_control)))
#up_control  <- ci_output[[1]]$est+1.96*sqrt(ci_output[[1]]$var)
#up_control [up_control < 0]  <- 0
#print (paste("up_control:", toString(up_control)))


## IMPORTANT !!!! - REMEMBER TO CHANGE the index number of case to 2 from 1 when you implement groups (case vs control)

low_case <- ci_output[[1]]$est-1.96*sqrt(ci_output[[1]]$var)
low_case [low_case < 0]  <- 0
print (paste("low_case:", toString(low_case)))
up_case  <- ci_output[[1]]$est+1.96*sqrt(ci_output[[1]]$var)
up_case [up_case < 0]  <- 0
print (paste("up_case:", toString(up_case)))
